components/support/sql/src/lazy.rs

```diff
@@ -0,0 +1,141 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use crate::open_database::{open_database_with_flags, ConnectionInitializer, Error};
+use interrupt_support::{register_interrupt, SqlInterruptHandle, SqlInterruptScope};
+use parking_lot::{MappedMutexGuard, Mutex, MutexGuard};
+use rusqlite::{Connection, OpenFlags};
+use std::{
+    path::{Path, PathBuf},
+    sync::{Arc, Weak},
+};
+
+/// Lazily-loaded database with interruption support
+///
+/// In addition to the [Self::interrupt] method, LazyDb also calls
+/// [interrupt_support::register_interrupt] on any opened database.  This means that if
+/// [interrupt_support::shutdown] is called it will interrupt this database if it's open and
+/// in-use.
+pub struct LazyDb<CI> {
+    path: PathBuf,
+    open_flags: OpenFlags,
+    connection_initializer: CI,
+    connection: Mutex<Option<Connection>>,
+    // It's important to use a separate mutex for the interrupt handle, since the whole point is to
+    // interrupt while another thread holds the connection mutex.  Since the only mutation is
+    // setting/unsetting the Arc, maybe this could be sped up by using something like
+    // `arc_swap::ArcSwap`, but that seems like overkill for our purposes. This mutex should rarely
+    // be contested and interrupt operations execute quickly.
+    interrupt_handle: Mutex<Option<Arc<SqlInterruptHandle>>>,
+}
+
+impl<CI: ConnectionInitializer> LazyDb<CI> {
+    /// Create a new LazyDb
+    ///
+    /// This does not open the connection and is non-blocking
+    pub fn new(path: &Path, open_flags: OpenFlags, connection_initializer: CI) -> Self {
+        Self {
+            path: path.to_owned(),
+            open_flags,
+            connection_initializer,
+            connection: Mutex::new(None),
+            interrupt_handle: Mutex::new(None),
+        }
+    }
+
+    /// Lock the database mutex and get a connection and interrupt scope.
+    ///
+    /// If the connection is closed, it will be opened.
+    pub fn lock(&self) -> Result<(MappedMutexGuard<'_, Connection>, SqlInterruptScope), Error> {
```

Comment:
Calling `lock()` twice (or any other method, i.e. `close()`, that requires a lock on `self.connection` ) from the same thread would cause a dead lock unless the `MappedMutexGuard` gets dropped before the 2nd call. We might want to note that in the docstring to avoid foot-guns.

Reply:
Looks good, thanks for adding that!

Assuming that the lock will eventually be released, I think calling `lock()` and `close()` from two separate threads should be fine, correct? Calling them from the same thread would definitely be problematic.